pkg/tfbridge/property_path.go

```diff
@@ -58,6 +63,119 @@ func lookupSchemas(
 	return LookupSchemas(schemaPath, tfs, ps)
 }
 
+func combinePropertyMapKeys(
+	object1, object2 resource.PropertyMap,
+) map[resource.PropertyKey]struct{} {
+	combined := make(map[resource.PropertyKey]struct{})
+	for k := range object1 {
+		combined[k] = struct{}{}
+	}
+	for k := range object2 {
+		combined[k] = struct{}{}
+	}
+	return combined
+}
+
+// SkipChildrenError is an error that can be returned by the visitor to skip the children of the current step.
+type SkipChildrenError struct{}
+
+func (SkipChildrenError) Error() string {
+	return "skip children"
+}
+
+type twoPropertyValueVisitor func(path propertyPath, val1, val2 resource.PropertyValue) error
+
+// walkTwoPropertyValues walks the two property values and calls the visitor for each step.
+// It returns an error if the visitor returns an error other than SkipChildrenError.
+//
+// The visitor can return SkipChildrenError to skip the children of the current step.
+// In case the two values have different types, we walk both values, starting with val1.
+func walkTwoPropertyValues(
+	path propertyPath,
+	val1, val2 resource.PropertyValue,
+	visitor twoPropertyValueVisitor,
+) error {
+	err := visitor(path, val1, val2)
+	if err != nil {
+		if errors.Is(err, SkipChildrenError{}) {
+			return nil
+		}
+		return err
+	}
+
+	if val1.IsArray() && val2.IsArray() {
+		arr1 := val1.ArrayValue()
+		arr2 := val2.ArrayValue()
+		for i := range max(len(arr1), len(arr2)) {
+			childPath := path.Index(i)
+			var childVal1, childVal2 resource.PropertyValue
+			if i >= len(arr1) {
+				childVal1 = resource.NewNullProperty()
+			} else {
+				childVal1 = arr1[i]
+			}
+			if i >= len(arr2) {
+				childVal2 = resource.NewNullProperty()
+			} else {
+				childVal2 = arr2[i]
+			}
+			err := walkTwoPropertyValues(childPath, childVal1, childVal2, visitor)
+			if err != nil {
+				return err
+			}
+		}
+	} else if val1.IsObject() && val2.IsObject() {
+		obj1 := val1.ObjectValue()
+		obj2 := val2.ObjectValue()
+		combined := combinePropertyMapKeys(obj1, obj2)
+		for k := range combined {
+			err := walkTwoPropertyValues(path.Subkey(k), obj1[k], obj2[k], visitor)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		if val1.IsArray() {
```

Comment:
Why are these clauses needed, curious? 

This can possibly made to work both ways, but here you have a collection that's mismatching on type, perhaps it is a scalar the wrong collection, but the walker continues to walk down this collection knowing 100% it's going to match null at the other end.

It's like:

```
walk(val1, x => visit(x, nil))
walk(val2, x => visit(nil, x))
```

I guess there's a certain symmetry of doing this as opposed to bailing, as every sub-element of X and Y gets visited by the zip visitor, it's like an outer join.

Reply:
Thanks for the suggestion, I've changed this to not recurse into nil collections.